### DISCUSSION TITLE
Deploy

### DISCUSSION THREAD

**Discussion by zferdz** [idom-team/idom#1098] [Question] [answered]
Is there any guide or even possible to deploy on:

* GCP (Google Coud platform)
* Netlify
* Space
* Github
* Heroku
* Others

I have already [read](https://reactpy.dev/docs/guides/getting-started/running-reactpy.html#) this documentation.

Thanks.

**Comment by Archmonger** [comment]
Deployment instructions are going to be based on your chosen backend.

For example, if you use the Flask or Django backend and want to deploy via Azure, a tutorial like this would describe the process:
https://learn.microsoft.com/en-us/azure/app-service/tutorial-python-postgresql-app?tabs=flask%2Cwindows

**Reply by zferdz** [reply]
Thanks I will try that.

**Comment by zferdz** [comment]
I did try the fastapi in [deta-space](https://deta.space/) (not a reactpy) from this [guide](https://fastapi.tiangolo.com/deployment/deta/). It worked.

Then I tried the reactpy [sample code](https://reactpy.dev/docs/guides/getting-started/running-reactpy.html#) for fastapi or the code below.

*main.py*

```python
from reactpy import component, html
from reactpy.backend.fastapi import configure
from fastapi import FastAPI  # missing in official sample code.


@component
def HelloWorld():
    return html.h1("Hello, world!")


app = FastAPI()
configure(app, HelloWorld)
```

But I got nothing. There was no `Hello, world!` displayed.

The project myreactpy log in deta-space does not indicate any errors.

And this is my Spacefile for deta-space. Typical for fastapi use.

```
v: 0
micros:
  - name: myreactpy
    src: .
    engine: python3.9
    run: uvicorn main:app
```

**Reply by Archmonger** [reply]
Can you check your browser console logs? I'm willing to bet that a standard fast API configuration does not forward websockets.

ReactPy requires websockets to function.

**Reply by zferdz** [reply]
Right, I got "failed: Error during WebSocket handshake"

**Reply by Archmonger** [reply]
Yep you'll need to figure out where exactly you need to enable WebSocket support. If the same ReactPy configuration works in a local environment, then it's likely some configuration on your hosting provider.

If you are using Nginx, that also needs to be configured for websockets.

**Comment by zferdz** [comment]
> Deployment instructions are going to be based on your chosen backend.
> 
> For example, if you use the Flask or Django backend and want to deploy via Azure, a tutorial like this would describe the process: https://learn.microsoft.com/en-us/azure/app-service/tutorial-python-postgresql-app?tabs=flask%2Cwindows

I can't register in azure at the moment.

I will try gcp instead as I already have an account on it.

**Comment by rmorshea** [comment]
The interactive examples in the docs are backed by Heroku:
- [Dockerfile](https://github.com/reactive-python/reactpy/blob/main/docs/Dockerfile)
- [GH Action for Deployment](https://github.com/reactive-python/reactpy/blob/main/.github/workflows/deploy-docs.yml)

**Comment by zferdz** [comment]
I tried GCP, fastapi worked, but reactpy on fastapi backend did not work.

**Reply by rmorshea** [reply]
I suspect that [this](https://github.com/reactive-python/reactpy/issues/1041) is the main reason for the issues on GCP. The GCP docs state:

> WebSockets clients connecting to Cloud Run should handle reconnecting to the server if the request times out or the server disconnects. You can achieve this in browser-based clients by using libraries such as [reconnecting-websocket](https://github.com/joewalnes/reconnecting-websocket/) or by handling "disconnect" events if you are using the [SocketIO](https://socket.io/docs/v3/server-socket-instance#disconnect) library.

It is likely that our reconnection logic is insufficient. The GCP docs suggest using the same `reconnecting-websocket` library I did in the issue.

**Comment by fsmosca** [comment, marked as answer]
This is what I have tried so far and they worked.

## Deploy for production

### GCP (Google Cloud Platform)

- [x] reactpy with flask backend
See [rflask](https://github.com/fsmosca/rflask)
- [x] reactpy with fastapi backend
See [gcp-reactpy-fastapi](https://github.com/fsmosca/gcp-reactpy-fastapi)

**Reply by Archmonger** [reply]
Thanks for looking into this!

If you have time, it would be great if you could write tutorial(s) in GitHub Discussions "share" category!

**Reply by fsmosca** [reply]
I will write one day.

**Comment by yickysan** [comment]
I also found you can deploy reactpy using a docker image. It works for platforms that don't support websockets.
